course-content/21_getting_data_with_pandas.qmd

```diff
@@ -0,0 +1,427 @@
+
+```{python}
+series_a.astype('Int64')
+```
+:::
+
+We commonly want to make new series out of old ones inside our dataframes. Operations are typically done on an element by element basis. We will see many examples of these in future sessions as we learn to manipulate dataframes, but here is a short example of what we may wish to do:
+
+```{python}
+series_a = pd.Series([2, 4, 6])
+series_b = pd.Series([1, 2, 3])
+
+display(series_a + series_b)
+display(series_a > 3)
+display(series_a*5)
+```
+
+## Bringing in our own data
+
+Enough about series. Let's talk dataframes! This is the main tool in our pandas toolkit. As we showed earlier, it is simply a collection of series all stacked together like an excel spreadsheet. There are many different ways to create a dataframe including:
+
+ * Within python itself
+ * From a csv, excel file, or some other local tabular format
+ * From more exotic data sources, such as a parquet file, json, or a website
+ * From a SQL database
+ 
+For the purposes of this course, we are going to focus on opening up local datafiles (with the most common type being a `csv` or `excel` file), and then utilizing the data once it is in python. To bring in data from a csv or excel file, we utilize the pandas methods `read_csv()` or `read_excel()`, with the only required argument being the `path` to the datafile.
+
+But first we need some data! [Navigate to this URL](https://raw.githubusercontent.com/bcgov/ds-intro-to-python/main/data/gapfinder.csv), right click on the data, and save the csv as `gapfinder.csv` in a folder called `data` inside our project folder. Now that we have a dataset, let's load it into pandas. 
+
+:::{.callout-tip}
+## The Data Folder
+
+While everyone may organize their projects and folders slightly differently, there are some general principles to adhere to that make project management easier. Arguably the most important of these is to treat input data as **read only**. If you have an excel spreadsheet, it is tempting to go make changes to the data directly in the spreadsheet: I'll just tweak a single value here, or add a column there. However, once we start doing this we lose the concept of reproducibility! How did we get certain values if the changes are all hidden in the excel spreadsheet? By keeping data as read only and making all of our changes in the python scripts, the processing is not only reproducible but also transparent to others. 
+
+Because of this, it is common to have a folder exclusively for raw data. Secondary folders may then be set-up for 'cleaned' data, python scripts, analysis outputs and more. 
```

Comment:
I will have introduced the value of setting up and working with directories in 1.3 working with code so this will be reinforement

/docs
/data
/output
/src
etc.